File: sql_modeling/src/utils/fit_eula_pops.py
import pandas as pd
import numpy as np
from scipy.optimize import curve_fit
import sys
sys.path.append( "." )
import settings

# Read the CSV file
df = pd.read_csv(sys.argv[1])

# Function to fit a line (linear regression)
def linear_fit(x, m, b):
    return m * x + b

# Group data by 'node' and fit a line for each group
fits = {}
for node, group in df.groupby('node'):
    x_data = group['t'].values
    y_data = group['pop'].values

    # Perform linear regression
    params, _ = curve_fit(linear_fit, x_data, y_data)

    # Save the fit parameters to a dictionary
    fits[node] = params

# Save the fits dictionary to a NumPy file
np.save(settings.eula_pop_fits, fits)

File: jb/utils/fit_eula_pops.R
library(data.table)
library(nlstools)

settings <- new.env()
source("demographics_settings.R", local = settings)

# Function to perform linear fit
linear_fit <- function(x, m, b) {
  return(m * x + b)
}

# Read the CSV file
args <- commandArgs(trailingOnly = TRUE)
input_file <- args[1]
output_file <- settings$eula_pop_fits

df <- fread(input_file)

# Group data by 'node' and fit a line for each group
fits <- list()
nodes <- unique(df$node)

for (node in nodes) {
  group <- df[node == df$node, ]
  x_data <- group$t
  y_data <- group$pop
  
  # Perform linear regression
  model <- nls(y_data ~ linear_fit(x_data, m, b), start = list(m = 1, b = 0))
  params <- coef(model)
  
  # Save the fit parameters to a list
  fits[[as.character(node)]] <- params
}

# Save the fits list to an RDS file
saveRDS(fits, output_file)

cat("Fits saved to", output_file, "\n")

